libkbfs/rekey_fsm.go

```diff
@@ -0,0 +1,488 @@
+package libkbfs
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/keybase/client/go/logger"
+	"github.com/keybase/kbfs/tlf"
+)
+
+/*
+
+ This file defines a finite state machine (FSM) for rekey operation scheduling.
+ The state chart is described in following dot graph:
+
+digraph rekeyFSM {
+  graph [rankdir=LR]
+  start [shape=plaintext]
+
+  Idle -> Idle [label="*"]
+  Scheduled -> Scheduled [label="*"]
+  Started -> Started [label="*"]
+
+  start -> Idle
+  Idle -> Scheduled [label=Request]
+  Scheduled -> Scheduled [label=Request]
+  Scheduled -> Started [label=Timeup]
+  Started -> Scheduled [label="Finished(TTL valid && (rekey done || needs paper))"]
+  Started -> Idle [label="Finished (*)"]
+}
+
+*/
+
+// CtxRekeyTagKey is the type used for unique context tags within an
+// enqueued Rekey.
+type CtxRekeyTagKey int
+
+const (
+	// CtxRekeyIDKey is the type of the tag for unique operation IDs
+	// within an enqueued Rekey.
+	CtxRekeyIDKey CtxRekeyTagKey = iota
+)
+
+// CtxRekeyOpID is the display name for the unique operation
+// enqueued rekey ID tag.
+const CtxRekeyOpID = "REKEYID"
+
+type rekeyEventType int
+
+const (
+	_ rekeyEventType = iota
+	rekeyRequestEvent
+	rekeyFinishedEvent
+	rekeyTimeupEvent
+
+	rekeyShutdownEvent
+
+	rekeyKickoffEventForTest
+	rekeyCancelEventForTest
+)
+
+func (e rekeyEventType) String() string {
+	switch e {
+	case rekeyRequestEvent:
+		return "rekeyRequestEvent"
+	case rekeyFinishedEvent:
+		return "rekeyFinishedEvent"
+	case rekeyTimeupEvent:
+		return "rekeyTimeupEvent"
+	case rekeyShutdownEvent:
+		return "rekeyShutdownEvent"
+	case rekeyKickoffEventForTest:
+		return "rekeyKickoffEventForTest"
+	case rekeyCancelEventForTest:
+		return "rekeyCancelEventForTest"
+	default:
+		return "unknown"
+	}
+}
+
+// rekeyTask describes a rekey task.
+type rekeyTask struct {
+	// timeout, if non-nil, causes rekey to fail if it takes more than this
+	// duration since it enters rekeyStateStarted.
+	timeout     *time.Duration
+	ttl         int
+	promptPaper bool
+
+	ctx context.Context
```

Comment:
We definitely want a `ctx` threaded throughout the rekey, I just don't want it saved to a struct if possible.  But now that I've looked through all the code I don't see a great way to do that.  I suppose adding a `rekeyState.getContext` functor might be a little better (that is passed in explicitly to `reactToEvent`), since then at least it's clear that the context is not supposed to be modified by the FSM.